lib/services/current_track_metadata_provider.dart

```diff
@@ -0,0 +1,42 @@
+import 'dart:async';
+
+import 'package:flutter/material.dart';
+import 'package:flutter_riverpod/flutter_riverpod.dart';
+import 'package:get_it/get_it.dart';
+
+import 'package:finamp/models/finamp_models.dart';
+import 'package:finamp/models/jellyfin_models.dart';
+import 'package:finamp/services/queue_service.dart';
+import 'metadata_provider.dart';
+
+/// Provider to handle pre-fetching metadata for upcoming tracks
+final currentTrackMetadataProvider =
+    AutoDisposeFutureProvider<MetadataProvider?>((ref) async {
+  final List<FinampQueueItem> precacheItems =
+      GetIt.instance<QueueService>().peekQueue(next: 3, previous: 1);
+  for (final itemToPrecache in precacheItems) {
+    BaseItemDto? base = itemToPrecache.baseItem;
+    if (base != null) {
+      // only fetch lyrics for the current track
+      final request = MetadataRequest(item: base, includeLyrics: true);
+      unawaited(ref.watch(metadataProvider(request).future));
+    }
+  }
+
+  final currentTrack = ref.watch(currentSongProvider).value?.baseItem;
+  if (currentTrack != null) {
+    final request = MetadataRequest(
+      item: currentTrack,
+      includeLyrics: true,
+    );
+    var out = ref.watch(metadataProvider(request)).valueOrNull;
```

Comment:
I think 
```
return ref.watch(metadataProvider(request).future);
```
is actually better, and will properly reflect the underlying loading state.  And then you can just stick an unconditional ref.keepAlive() into this function so that we don't need a listener in the now playing bar.

Reply:
thanks for pointing that out, I'm now doing that. I definitely was tired when implementing the loading state enum 😂  

I also changed the passive listen in the now playing bar to a call to closing the `KeepAliveLink` of the current track metadata provider, which should make sure that settings changes are properly applied (in case the metadata can be customized in the future), and it's possible to recover from failures.  
But the main reason for this was properly re-fetching the lyrics when toggling offline mode, which now seems to work fine.

Reply:
I don't think I fully though this through either.  Keepalive isn't going to get you you prefetching.  I checked the docs and the suggested method is wrapping everything in a consumer that watches the ref.  You can give consumers a child outside their build methods to prevent rebuilding everything unnecessarily. Also, as an alternative to the keepalive link, there's a provider version of the finamp settings you can watch to autodispose when it changes without needing the parents to worry about it.  You can also use select to specifically only rebuild when offline changes and not other attributes, because I there's a couple settings like current library that you really don't need to dispose for.

Reply:
Okay, I've pushed another change. I went with a ref.listen instead of the ref.watch outside the build method, not sure if that's equivalent? Watching and selecting the settings seems to work as well.  

However, now the prefetching/caching doesn't seem to work at all. I noticed that the was I was using the provider, it gave me the previous value if available, instead of showing the loading state. I've fixed that now, but the loading state is now briefly shown with every track change. Not sure if that's how it's supposed to we working and my loading state logic just has to be changed, or if the prefetching is actually not working...

Reply:
If you're talking about the super quick blinks, that's async lag, becuase the async current metadata provider doesn't run until after the build method that calls it has completed..  You can get around it by making the current metadata provider an AutoDisposeProvider<AsyncValue<MetadataProvider?>>.  Then the provider can syncronously hand over the child value without adding an unneeded async cycle.

Reply:
I think the delay is async lag.  You should be able to convert the current metadata provider into an AutodisposeProvider<Asyncvalue<MetadataProvider?>> to work around this.

Reply:
worked like a charm! But I have idea why :D

Reply:
Basically, the problem is that the current provider was a future provider that ran an async function to build.  Async code blocks return futures, and futures cannot be evaluated synchronously, ever.  This is a problem because the entire build sequence for a frame is synchronous, which means no futures can evaluate until it is fully complete.  This is why future providers return asyncvalue, which either synchronously has the future’s result already, or synchronously doesn’t.  So after the current provider invalidated we had to build the widget twice.  On the first build we trigger the provider build, but it cannot evaluate the returned future and provides a loading async value.  Then during the async gap after the widget tree build completes, the future is evaluated and the provider updates to a data asyncvalue.  This triggers the second widget build with the final data on the next frame.  The new code synchronously returns the chosen metadataProvider’s asyncvalue, which should already be the final data AsyncValue due to the precaching.

Reply:
Yeah, that's pretty much it.